src/libNetwork/RumorManager.cpp

```diff
@@ -0,0 +1,152 @@
+#include "RumorManager.h"
+
+#include <chrono>
+#include <thread>
+
+// TODO: would be nice to decouple from this header
+#include "P2PComm.h"
+#include "common/Messages.h"
+
+namespace
+{
+    // TODO: allow them to passed in as parameters
+    const std::chrono::milliseconds ROUND_TIME(500);
+
+    Message::Type convertType(uint32_t type)
+    {
+        switch (type)
+        {
+        case 1:
+            return Message::PUSH;
+        case 2:
+            return Message::PUSH;
+        default:
+            return Message::UNDEFINED;
+        }
+    }
+
+} // anonymous namespace
+
+// PRIVATE METHODS
+template<typename Container>
+void RumorManager::addRumorImp(const Container& peers,
+                               const RumorManager::RawBytes& message)
+{
+    std::lock_guard<std::mutex> guard(m_mutex); // critical section
+
+    int rumorId = m_rumorIdGenerator++;
+    m_rumorIdBimap.insert(RumorBimap::value_type(rumorId, message));
+
+    int peerIdGenerator = 0;
+    std::unordered_set<int> peerSet;
+    PeerBimap peerBimap;
+    for (const auto& p : peers)
+    {
+        peerBimap.insert(PeerBimap::value_type(peerIdGenerator++, p));
+        peerSet.insert(peerIdGenerator++);
+    }
+
+    m_rumorState.insert({rumorId, {RumorMember(peerSet), peerBimap}});
+
+    scheduleRounds(rumorId);
+}
+
+void RumorManager::scheduleRounds(int rumorId)
+{
+    // critical section
+    auto it = m_rumorState.find(rumorId);
+    if (it == m_rumorState.end())
+    {
+        return;
+    }
+    const RumorMember& member = it->second.m_member;
+    int numOfRounds = member.networkConfig().maxRoundsTotal();
+
+    std::thread([&]() {
```

Comment:
pass numOfRounds by copy (and extract this in another method for better readability)